clk: ad9545: Add support

tachicialex opened the pull request:
Add support for AD9545 Quad Input, 10-Output, Dual DPLL/IEEE 1588,
1 pps Synchronizer and Jitter Cleaner.

Driver currently implements:
- reference validation
- phase build-out mode only
- set_rate/get_rate/set_phase/get_phase on all 10 clock outputs

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+#define AD9545_NCO_MAX_FREQ		65535
+
+static const unsigned int ad9545_apll_rate_ranges_hz[2][2] = {
+	{2400000000U, 3200000000U}, {3200000000U, 4000000000U}
```
curious: why the U suffix. Isn't that implied by the type of the table?

tachicialex replied:
added that in order to get rid of: clk-ad9545.c:160:2: warning: this decimal constant is unsigned only in ISO C90, is there any other way to avoid this warning?

commodo replied:
no there isn't;
here's an explanation about it [in the answer part]:
https://stackoverflow.com/questions/9941261/warning-this-decimal-constant-is-unsigned-only-in-iso-c90

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+			continue;
+
+		ret = fwnode_property_read_u32(child, "reg", &ref_ind);
+		if (ret < 0)
```
maybe a `dev_err()` message would be helpful here

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+
+			st->ref_in_clks[ref_ind].s_conf = val;
+		}
+
```
This gives the impression that it can be single ended and differential at the same time... Also, do we need both properties or do we have a default mode? If there's a default mode I would only add one boolean property for the other mode

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+
+		ret = fwnode_property_read_u32(child, "adi,ref-dtol-pbb", &val);
+		if (ret < 0)
+			return ret;
```
nit: new line here

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+				st->ref_in_clks[ref_ind].monitor_hyst_scale = i;
+				break;
+			}
+		}
```
nit: new line here

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+	ret = regmap_read(st->regmap, AD9545_PLL_STATUS, &val);
+	for (i = 0; i < 2; i++)
+		if (st->pll_clks[i].pll_used && !AD9545_PLLX_LOCK(i, val))
+			dev_err(st->dev, "PLL%d unlocked.\n", i);
```
nit: If we are not going to return error, I would change it to `dev_warn()`

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+	u32 reg;
+	int i;
+
+	for (i = 0; i < 2; i++) {
```
`ARRAY_SIZE(st->pll_clks)`?

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+				return ret;
+
+			cal_count += 1;
+			mdelay(100);
```
Any reason for `mdelay()`. Do we need the sleep to be atomic? Most cases we want `msleep()`

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+
+			return ad9545_io_update(st);
+		}
+	}
```
I'm bit puzzled by this. Why do we need two loops? I don't see `i` being used anywhere... Or is it that we try at most 2 times?

tachicialex replied:
yeah, datasheet specifies calibration should be tried twice. Weird state machine at Figure 121. AD9545 System Clock Initialization Subprocess.

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+	int ret;
+	int i;
+
+	for (i = 0; i < 2; i++) {
```
ARRAY_SIZE() instead of plain 2?

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+
+	ret = regmap_bulk_read(clk->st->regmap, AD9545_DPLLX_N_DIV(clk->address), &regval, 4);
+	if (ret < 0)
+		return ret;
```
nit: new line here

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+	regval = 0;
+	ret = regmap_bulk_read(clk->st->regmap, AD9545_DPLLX_MOD_DIV(clk->address), &regval, 3);
+	if (ret < 0)
+		return ret;
```
nit: new line

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+	regval = 0;
+	ret = regmap_bulk_read(clk->st->regmap, AD9545_DPLLX_FRAC_DIV(clk->address), &regval, 3);
+	if (ret < 0)
+		return ret;
```
nit: new line

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+	}
+
+	st->onecell.clk_num = ARRAY_SIZE(ad9545_out_clk_names);
+	st->onecell.clks = devm_kzalloc(st->dev, 10 * sizeof(*st->onecell.clks), GFP_KERNEL);
```
error check?

commodo commented:
Can you make the driver Dual licensed?
https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=213&view=logs&j=330111c3-2990-58a5-de10-4a04d1985069&t=cd37190c-9ed1-50f4-4035-5eb954a4dfe6
[ thanks for testing this :P  ; i wasn't 100% sure it works ]

And also can you add it to the build so that it gets built?
https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=213&view=logs&j=1e15a8a0-9bb0-5d00-46e4-70ed8d4e47e6&t=a0ef7065-0132-5058-b2f3-a74338c90e11

nunojsa commented on Documentation/devicetree/bindings/clock/clk-ad9545.yaml:
```diff
@@ -0,0 +1,373 @@
+        allOf:
+          - $ref: /schemas/types.yaml#/definitions/uint32
+          - minimum: 1
+          - maximum: 1073741824
```
Did you run `make dt_binding_check`? I'm not sure this will work as it is. IIRC, this should be something like:

```
allOf:
   - $ref: /schemas/types.yaml#/definitions/uint32
   - minimum: 1
     maximum: 1073741824
```

Ditto for all the other properties like this

nunojsa commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1618 @@
+		}
+		if (i == ARRAY_SIZE(ad9545_hyst_scales_bp))
+			return -EINVAL;
+
```
So, in your scheme you have:

```
required:
  - compatible
  - reg
```

But it looks like that all the following properties are required? I guess all of these properties are not mandatory right? And if they are, they should be set as `required` in the schema. So, take a look on all of the supported properties and make sure to:

1. If the property is mandatory, fail the device `probe()` and maybe print some meaningful error message. And mark the property as required in the schema.
2.  If not mandatory, do not fail the device proble().

tachicialex commented:
Changelog V1 -> V2:
- added Dual BSD/GPL license
- i2c probe changed to probe_new
- in ad9545_parse_dt_inputs, if adi,single-ended-mode is not present
assume that adi,differential-mode is specified in dt.
- changed mdelay to msleep
- added ARRAY_SIZE in for loops instead of hard-coded values
- check return from devm_kzalloc
- added required properties to dt-bindings
- added to ADI Kconfig

commodo commented on drivers/clk/Kconfig.adi:
```diff
@@ -0,0 +1,8 @@
+config CLK_ALL_ADI_DRIVERS
+	tristate "Build all Analog Devices CLOCK Drivers"
+	select CLK_AD9545
+	select CLK_AD9545_I2C
```
symbols seem to be `COMMON_CLK_AD9545_I2C` and so on
this still doesn't build yet

commodo commented on drivers/clk/clk-ad9545/Makefile:
```diff
@@ -0,0 +1,9 @@
+#
+
+# When adding new entries keep the list in alphabetical order
+obj-$(CONFIG_COMMON_CLK_AD9545) += clk-ad9545.o
```
i'm not sure that doing a `clk-ad9545` folder is generally done;
maybe we can make a `drivers/clk/adi` folder;  NXP and TI seem to have one;
we have enough clock chips to fit into such a folder;

commodo commented on drivers/clk/clk-ad9545/clk-ad9545-i2c.c:
```diff
@@ -0,0 +1,63 @@
+ *
+ * Copyright 2020 Analog Devices Inc.
+ *
+ * Licensed under the Dual BSD/GPL.
```
remove `Licensed under the Dual BSD/GPL.` here

commodo commented on drivers/clk/clk-ad9545/clk-ad9545-spi.c:
```diff
@@ -0,0 +1,77 @@
+MODULE_DEVICE_TABLE(of, ad9545_spi_of_match);
+
+static const struct spi_device_id ad9545_spi_id[] = {
+	{"adi,ad9545-spi", 0},
```
spi device IDs don't need vendor prefix, nor `-spi` suffix;
`"ad9545"` is sufficient;

commodo commented on drivers/clk/clk-ad9545/clk-ad9545-spi.c:
```diff
@@ -0,0 +1,77 @@
+}
+
+static const struct of_device_id ad9545_spi_of_match[] = {
+	{ .compatible = "adi,ad9545-spi" },
```
this compat string can be simply  `adi,ad9545`

commodo commented on drivers/clk/clk-ad9545/clk-ad9545-i2c.c:
```diff
@@ -0,0 +1,63 @@
+MODULE_DEVICE_TABLE(of, ad9545_i2c_of_match);
+
+static const struct i2c_device_id ad9545_i2c_id[] = {
+	{"adi,ad9545-i2c"},
```
also, this needs to be just `ad9545`

commodo commented on drivers/clk/clk-ad9545/clk-ad9545-i2c.c:
```diff
@@ -0,0 +1,63 @@
+}
+
+static const struct of_device_id ad9545_i2c_of_match[] = {
+	{ .compatible = "adi,ad9545-i2c" },
```
and this `adi,ad9545`

commodo commented on include/dt-bindings/clock/ad9545.h:
```diff
@@ -0,0 +1,35 @@
+ *
+ * Copyright 2020 Analog Devices Inc.
+ *
+ * Licensed under the GPL-2.
```
remove this license `Licensed under the GPL-2.` 
and maybe update the `SPDX-License-Identifier: GPL-2.0 */` to include BSD

commodo commented on drivers/clk/clk-ad9545/clk-ad9545.c:
```diff
@@ -0,0 +1,1629 @@
+ *
+ * Copyright 2020 Analog Devices Inc.
+ *
+ * Licensed under the Dual BSD/GPL.
```
remove `Licensed under the Dual BSD/GPL.`

tachicialex commented:
Changelog V2 -> V3:
- renamed clk-ad9545 dir to adi
- removed -spi -i2c from compatibles strings
- removed " Licensed under the Dual BSD/GPL" comment from all files.
- fixed misspelled kconfig symbols in drivers/clk/Kconfig.adi

tachicialex commented:
Changelog V3 -> V4:
- add output clock gating

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+	if (ret < 0)
+		return ret;
+
+	ret = ad9545_setup(st);
```
`return ad9545_setup()`

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+		/* APLL VCO calibration operation */
+		cal_count = 0;
+		while (cal_count < 2) {
+			ret = regmap_write(st->regmap, AD9545_PWR_CALIB_CHX(i), 0);
```
the block inside the while() could be in a function called like  `ad9545_calib_apll(st, i)`

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+		if (ret < 0)
+			return ret;
+
+		rate = clk_round_rate(nco->hw.clk, nco->nco_freq_hz);
```
the clock framework may have some facilities to set these frequencies using `assigned-clocks` and `assigned-clock-rates`
an example would be:
```
                fmc_mii_clk1: clock-generator@32 {
                        #clock-cells = <0>;
                        compatible = "silabs,si514";
                        reg = <0x32>;
                        clock-output-names = "si514_fmc_mii_clk1";
                        assigned-clocks = <&fmc_mii_clk1>;
                        assigned-clock-rates = <50000000>;
                };
```

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+		if (ret < 0)
+			return ret;
+
+		rate = clk_round_rate(pll->hw.clk, pll->out_freq_hz);
```
same here with `assigned-clocks` and `assigned-clock-rates`

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+	u32 m;
+	u32 n;
+
+	freq = ad9545_calc_pll_params(clk, rate, *parent_rate, &m, &n, &frac, &mod);
```
`return ad9545_calc_pll_params()`?

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+	output_rate = mul_u64_u32_div(*frac * parent_rate, m_div, *mod);
+	output_rate += parent_rate * dpll_n_div * m_div;
+
+	return DIV_ROUND_CLOSEST(output_rate, 2);
```
can the return value here be greater than `UINT32_MAX`?
if no, i would convert it back to u32;
if yes, maybe we can print a warning?

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+			reg |= (1 << i);
+	}
+
+	ret = regmap_write(st->regmap, AD9545_POWER_DOWN_REF, reg);
```
`return regmap_write()` ?

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+
+		st->onecell.clks[i] = st->out_clks[i].hw.clk;
+
+		rate = clk_round_rate(st->out_clks[i].hw.clk, st->out_clks[i].init_freq_hz);
```
i'm wondering here as well about `assigned-clock-rates` prop

commodo commented on drivers/clk/adi/clk-ad9545.c:
```diff
@@ -0,0 +1,1627 @@
+		if (ret < 0)
+			return ret;
+
+		ret = clk_set_phase(st->out_clks[i].hw.clk, st->out_clks[i].init_phase);
```
not sure if there is a `assigned-clock-phase` value; if not, we can maybe propose to the clk framework if it makes sense

commodo reviewed (approved):
a few notes
looks good

tachicialex commented:
Changelog V4 -> V5:
- moved while block form  ad9545_calib_aplls to ad9545_calib_apll
- removed init-freq/init-phase from dt bindings (for clock phase there is no
dt binding yet)
- added twocell_get callback with of_clk_add_provider:
<&ad9545_node type address> Where:
	- type is AD9545_CLK_OUT for outputs AD9545_CLK_PLL is for PLLs
	and AD9545_CLK_NCO is for auxiliary NCOs
	- address: ex: for output can be AD9545_Q0A -> AD9545_Q1BB
	- now all frequencies can be set via assigned-clock-rates
